EIPS/eip-4750.md

```diff
@@ -0,0 +1,185 @@
+4. Pops nothing and pushes nothing to data stack.
+5. Pushes to return stack an item:
+```
+(code_section_index = current_section_index, 
+offset = PC_post_instruction,
+stack_height = data_stack.height - types[code_section_index].inputs)
+```
+
+Under `PC_post_instruction` we mean the PC position after the entire immediate argument of `CALLF`. Data stack height is saved as it was before function inputs were pushed.
+
+> Note: code validation rules of EIP-3670 guarantee there is always an instruction following `CALLF` (since terminating instruction is required to be final one in the section), therefore `PC_post_instruction` always points to an instruction inside section bounds.
+
+6. Sets `current_section_index` to `code_section_index` and `PC` to `0`, and execution continues in the called section.
+
+#### `RETF`
+
+1. Does not have immediate arguments.
+2. If data stack has less than `caller_stack_height + types[code_section_index].outputs`, execution results in exceptional halt.
+3. Pops nothing and pushes nothing to data stack.
+4. Pops an item from return stack and sets `current_section_index` and `PC` to values from this item.
+    4.1. If return stack is empty after this, execution halts with success.
+
+### Code Validation
+
+In addition to container format validation rules above, we extend code section validation rules (as defined in [EIP-3670](./eip-3670.md)).
+
+1. Code validation rules of EIP-3670 are applied to every code section.
+2. EIP-3670 requires that immediate arguments of instructions (`PUSHn`, `RJUMP`, `RJUMPI`) are inside code section bounds. This rule is extended to include `CALLF` into the list of such instructions. That is, 2 bytes directly following `CALLF` must be inside code section.
+
+> TODO In case EIP-3670 includes terminator instruction requirement:
+> 3. List of allowed terminating instructions is extended to include `RETF`.
+> (rule 2 above is not really needed then, is is implied from this)
+3. Code section is invalid in case an immediate argument of any `CALLF` is greater than or equal to the total number of code sections.
+4. `RJUMP` and `RJUMPI` immediate argument value (jump destination relative offset) validation:
+    4.1. Code section is invalid in case offset points to a position outside of section bounds.
+    4.2. Code section is invalid in case offset points to one of two bytes directly following `CALLF` instruction.
+
+### Execution
+
+1. Execution starts at the first byte of the first code section, and PC is set to 0.
+2. Return stack is initialized to contain one item: `(code_section_index = 0, offset = 0, stack_height = 0)`
+3. Jumps destinations are allowed only to be inside current code section. `JUMP` and `JUMPI` result in exceptional halt when destination is outside of current section bounds.
+4. If any instruction would change the stack to height below `caller_stack_height`, execution results in exceptional halt. This rule replaces the old stack underflow check.
```

Comment:
```suggestion
4. If any instruction would change the stack to height below `caller_stack_height` or accesses a data stack item below that height, execution results in exceptional halt. This rule replaces the old stack underflow check.
```

Reply:
Explanation: It can be argued if `swap` instructions actually change the stack height.

Reply:
Good point!

Reply:
Changed.